Remove redundant union check in `KnownPanicsLint` const prop

gurry opened the pull request:
Removes the below check which prevents unions from being const propagated:https://github.com/rust-lang/rust/blob/f9dca46218d4b8efa062aec4fd0820cbb4942aa2/compiler/rustc_mir_transform/src/known_panics_lint.rs#L587-L594

It is not needed because after PR #124504 we mark unions as `NoPropagation` over here: https://github.com/rust-lang/rust/blob/f9dca46218d4b8efa062aec4fd0820cbb4942aa2/compiler/rustc_mir_transform/src/known_panics_lint.rs#L899-L902 which is enough to prevent them from being const propagated.

gurry commented on compiler/rustc_mir_transform/src/known_panics_lint.rs:
```diff
@@ -897,8 +885,8 @@ impl CanConstProp {
-                // Do not const prop unions as they can
-                // ICE during layout calc
+                // Unions are incompatible with const prop because
+                // Rust has no conpcept of the active variant of a union
```
Also improved this comment while I was at it. The new comment provides a more fundamental reason why we can't const prop unions

oli-obk replied:
Typo: conpcept

gurry replied:
Whoops

gurry replied:
Fixed

oli-obk commented on compiler/rustc_mir_transform/src/known_panics_lint.rs:
```diff
@@ -897,8 +885,8 @@ impl CanConstProp {
             if ty.is_union() {
-                // Do not const prop unions as they can
-                // ICE during layout calc
+                // Unions are incompatible with const prop because
```
Incompatible with this simple const prop impl

It's not a fundamental issue with const propagators

gurry replied:
Yeah, I meant that. Should've been more explicit. Will update it.

gurry replied:
Done

gurry committed:
Remove redundant union check in `KnownPanicsLint const prop

because we are already marking unions `NoPropagation` in
`CanConstProp::check()`. That is enough to prevent any attempts
at const propagating unions and this second check is not needed.

Also improve a comment in `CanConstProp::check()`

oli-obk commented:
Thanks!

@bors r+ rollup

bors commented:
:pushpin: Commit 741d40f327291404172b3f35377d8464688c2fe7 has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk 741d40f327291404172b3f35377d8464688c2fe7 -->
<!-- homu: {"type":"Approved","sha":"741d40f327291404172b3f35377d8464688c2fe7","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
